src/olympia/discovery/tests/test_admin.py

```diff
@@ -1226,3 +1277,58 @@ def test_can_not_delete_without_discovery_edit_permission(self):
         assert response.status_code == 403
         assert Shelf.objects.filter(pk=item.pk).exists()
         assert item.title == 'Recommended extensions'
+
+    def test_clean_criteria_search(self):
+        form = ShelfForm({
+            'title': "Recommended extensions",
+            'shelf_type': 'extension',
+            'criteria': '?recommended=true&sort=random&type=extension'})
+        assert form.is_valid(), form.errors
+        assert form.cleaned_data['criteria'] == (
+            '?recommended=true&sort=random&type=extension')
+
+    def test_clean_criteria_categories(self):
+        form = ShelfForm({
+            'title': "Alerts & Updates (Categories)",
+            'shelf_type': 'categories',
+            'criteria': '?slug=alerts-updates'})
+        assert form.is_valid(), form.errors
+        assert form.cleaned_data['criteria'] == '?slug=alerts-updates'
+
+    def test_clean_criteria_recommendations(self):
+        form = ShelfForm({
+            'title': "Recommended Add-ons",
+            'shelf_type': 'recommendations',
+            'criteria': '?recommended=true'})
+        assert form.is_valid(), form.errors
+        assert form.cleaned_data['criteria'] == '?recommended=true'
+
+    def test_clean_criteria_returns_404(self):
+        with self.assertRaises(Exception) as context:
+            form = ShelfForm({
+                'title': "Popular themes",
+                'shelf_type': 'theme',
+                'criteria': 'sort=users&type=statictheme'})
+            assert form.is_valid()
+            self.assertTrue(
+                'No data found - check criteria' in str(context.exception))
```

Comment:
`assert 'No data found - check criteria' in str(context.exception)` though you should be able to do `assert 'No data found - check criteria' in context.exception.message`, right?

Reply:
same for the following tests